src/Fixer/PhpUnit/PhpUnitTestAnnotationFixer.php

```diff
@@ -0,0 +1,572 @@
+        if (' ' === $line[$i]) {
+            ++$i;
+        }
+        $line = array_slice($line, $i);
+        $line = implode($line);
+
+        return $line;
+    }
+
+    /**
+     * Updates the depends tag on the current doc block.
+     *
+     * @param Line $line
+     *
+     * @return Line
+     */
+    private function updateDepends(Line $line)
+    {
+        if ('annotation' === $this->configuration['style']) {
+            return $this->removeTestFromDepends($line);
+        }
+
+        return $this->addtestToDepends($line);
+    }
+
+    /**
+     * @param Line $line
+     *
+     * @return Line
+     */
+    private function removeTestFromDepends(Line $line)
+    {
+        $line = \str_split($line->getContent());
+
+        $dependsIndex = $this->findWhereDependsFunctionNameStarts($line);
+        $dependsFunctionName = implode(array_slice($line, $dependsIndex));
+
+        if ($this->startsWith('test', $dependsFunctionName)) {
+            $dependsFunctionName = $this->removeTestFromFunctionName($dependsFunctionName);
+        }
+        array_splice($line, $dependsIndex);
+
+        return new Line(implode($line).$dependsFunctionName);
+    }
+
+    /**
+     * @param Line $line
+     *
+     * @return Line
+     */
+    private function addTestToDepends(Line $line)
+    {
+        $line = \str_split($line->getContent());
+        $dependsIndex = $this->findWhereDependsFunctionNameStarts($line);
+        $dependsFunctionName = implode(array_slice($line, $dependsIndex));
+
+        if (!$this->startsWith('test', $dependsFunctionName)) {
+            $dependsFunctionName = $this->addTestToFunctionName($dependsFunctionName);
+        }
+
+        array_splice($line, $dependsIndex);
+
+        return new Line(implode($line).$dependsFunctionName);
+    }
+
+    /**
+     * Helps to find where the function name in the doc block starts.
+     *
+     * @param array $line
+     *
+     * @return int
+     */
+    private function findWhereDependsFunctionNameStarts(array $line)
+    {
+        $counter = \count($line);
+
+        do {
+            --$counter;
+        } while (' ' !== $line[$counter]);
+
+        return $counter + 1;
+    }
+
+    /**
+     * @param Line[] $lines
+     * @param Tokens $tokens
+     * @param int    $docBlockIndex
+     *
+     * @return Line[]
+     */
+    private function addTestToDocBlockIfNeeded($lines, Tokens $tokens, $docBlockIndex)
+    {
+        $doc = new DocBlock($tokens[$docBlockIndex]->getContent());
+
+        if (!$this->doesDocBlockContainTest($doc)) {
+            $originalIndent = $this->detectIndent($tokens, $docBlockIndex);
+            $lineEnd = $this->whitespacesConfig->getLineEnding();
+
+            array_splice($lines, 1, 0, $originalIndent.' * @test'.$lineEnd);
+        }
+
+        return $lines;
+    }
+
+    /**
+     * @param DocBlock $doc
+     *
+     * @return bool
+     */
+    private function doesDocBlockContainTest(DocBlock $doc)
```

Comment:
can we just get annotation from docblock? it has all the logic already

Reply:
ping @BackEndTea

Reply:
I'm sorry, i don't understand exactly what it is you are asking.

Reply:
```php
private function doesDocBlockContainTest(DocBlock $doc)
{
    return !empty($doc->getAnnotationsOfType('test'));
}
```

Reply:
Then we get some issues with functions annotated like this:
```php
/**
 * This is a @test
 */
```
Which is a valid phpunit test

Reply:
It causes the ```Annotation does not get added when it is already present in a weird place``` case to break.

If we don't want to support that, i'm okay with that too

Reply:
as said in https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/3275#issuecomment-349796720 , we shouldn't handle that scenario

Reply:
👍 I'll update it then

Reply:
done